inst/include/cpp11/as.hpp

```diff
@@ -77,7 +77,12 @@ using is_enum = typename std::enable_if<std::is_enum<E>::value, E>::type;
 template <typename E>
 is_enum<E> as_cpp(SEXP from) {
   if (Rf_isInteger(from)) {
-    return static_cast<E>(as_cpp<typename std::underlying_type<E>::type>(from));
+    using Type = typename std::conditional<
+      std::is_same<char, typename std::underlying_type<E>::type>::value,
+      int,
+      typename std::underlying_type<E>::type
+    >::type;
+    return static_cast<E>(as_cpp<Type>(from));
```

Comment:
Maybe it'd be better to just use `intmax_t` unconditionally:
```suggestion
    return static_cast<E>(as_cpp<intmax_t>(from));
```

Reply:
I guess the other issue is enums could theoretically be signed or unsigned (https://stackoverflow.com/questions/33304874/is-there-an-integral-type-you-can-use-that-holds-any-enum-value)

Reply:
Hmm, you're right; we could have an enum like
```c++
enum : uintmax_t { zero = 0, max = numeric_limits<uintmax_t>::max() };
```
... and the behavior of my suggestion is implementation defined until c++20. Could we just make the `char` special case clear with a comment:
```suggestion
    using underlying_type = typename std::underlying_type<E>::type;
    using int_type = typename std::conditional<
      std::is_same<char, underlying_type>::value,
      int, // as_cpp<char> would trigger undesired string conversions
      underlying_type
    >::type;
    return static_cast<E>(as_cpp<int_type>(from));
```

Reply:
Scratch that: we don't define `as_cpp<char>()` and we should. Could you see if adding
```c++
template <typename T>
enable_if_char<T, SEXP> as_sexp(T from) {
  return safe[Rf_ScalarInteger](from);
}
```
resolves the issue?

Reply:
There actually is a `as_cpp<>` for `char`: 

```cpp
template <typename T>
using is_char =
    typename std::enable_if<std::is_same<typename std::decay<T>::type, char>::value,
                            typename std::decay<T>::type>::type;
template <typename T>
is_char<T> as_cpp(SEXP from) {
  if (Rf_isString(from)) {
    if (Rf_xlength(from) == 1) {
      const char* c_p = nullptr;
      unwind_protect([&] { c_p = Rf_translateCharUTF8(STRING_ELT(from, 0)); });
      return c_p[0];
    }
  }
  stop("Expected string vector of length 1");

  return T();
}
```

such that: 

``` r
cpp11::cpp_source(code = '
#include <cpp11.hpp>

enum class Color : char { red, green = 20, blue };

[[cpp11::register]]
int fun(SEXP value) {
  auto x = cpp11::as_cpp<char>(value);
  return static_cast<int>(x);
}

', quiet = FALSE)
#> clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I/Users/romainfrancois/.R/library/4.0/cpp11/include  -I/usr/local/include   -fPIC  -Wall -O3 -Wall -Wimplicit-int-float-conversion -c /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/Rtmpg1dGCP/filed45fbb133fe/src/code_0.cpp -o /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/Rtmpg1dGCP/filed45fbb133fe/src/code_0.o
#> clang++ -mmacosx-version-min=10.13 -std=gnu++11 -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -I/Users/romainfrancois/.R/library/4.0/cpp11/include  -I/usr/local/include   -fPIC  -Wall -O3 -Wall -Wimplicit-int-float-conversion -c /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/Rtmpg1dGCP/filed45fbb133fe/src/cpp11.cpp -o /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/Rtmpg1dGCP/filed45fbb133fe/src/cpp11.o
#> clang++ -mmacosx-version-min=10.13 -std=gnu++11 -dynamiclib -Wl,-headerpad_max_install_names -undefined dynamic_lookup -single_module -multiply_defined suppress -L/Library/Frameworks/R.framework/Resources/lib -L/usr/local/lib -o /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/Rtmpg1dGCP/filed45fbb133fe/src/code_0.so /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/Rtmpg1dGCP/filed45fbb133fe/src/code_0.o /private/var/folders/4b/hn4fq98s6810s4ccv2f9hm2h0000gn/T/Rtmpg1dGCP/filed45fbb133fe/src/cpp11.o -F/Library/Frameworks/R.framework/.. -framework R -Wl,-framework -Wl,CoreFoundation
fun("a")
#> [1] 97
```

<sup>Created on 2020-08-06 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

Reply:
it's clearer with the comment, thanks @bkietz